experiments/dsl/src/main/scala/org/apache/gearpump/streaming/flatmap/Stream.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.flatmap
+
+import org.apache.gearpump.streaming.flatmap.op._
+import org.apache.gearpump.streaming.task.Task
+import org.apache.gearpump.util.Graph
+import org.slf4j.LoggerFactory
+
+class Stream[T](dag: Graph[Op, OpEdge], private val root: Op) {
+
+  def flatMap[R](fun: T => TraversableOnce[R]): Stream[R] = {
+    val flatMapOp = FlatMapOp(fun)
+    dag.addVertex(flatMapOp )
+    dag.addEdge(root, Direct, flatMapOp)
+    new Stream(dag, flatMapOp)
+  }
+
+  def map[R](fun: T => R): Stream[R] = {
+    this.flatMap { data =>
+      Option(fun(data))
+    }
+  }
+
+  def filter(fun: T => Boolean): Stream[T] = {
+    this.flatMap { data =>
+      if (fun(data)) Option(data) else None
+    }
+  }
+
+  def reduce(fun: (T, T) => T): Stream[T] = {
+    val reduceOp = ReduceOp(fun)
+    dag.addVertex(reduceOp)
+    dag.addEdge(root, Direct, reduceOp)
+    new Stream(dag, reduceOp)
+  }
+
+  def log(): Unit = {
+    this.map(msg => LoggerFactory.getLogger("dsl").info(msg.toString))
+
+  }
+
+  def merge(other: Stream[T]): Stream[T] = {
+    val mergeOp = MergeOp(root, other.root)
+    dag.addVertex(mergeOp)
+    dag.addEdge(root, Direct, mergeOp)
+    dag.addEdge(other.root, Shuffle, mergeOp)
+    new Stream(dag, mergeOp)
+  }
+
+  def groupBy[R](fun: T => R, parallism: Int): Stream[T] = {
+    val groupOp = GroupByOp(fun, parallism)
+    dag.addVertex(groupOp)
+    dag.addEdge(root, Shuffle, groupOp)
+    new Stream(dag, groupOp)
+  }
+
+  def reParition(newPartitionNumber: Int): Stream[T] = {
+    this.groupBy(data => (data.hashCode & Integer.MAX_VALUE), newPartitionNumber)
+  }
+
+  def process[R](processor: Class[_ <: Task], parallism: Int): Stream[R] = {
+    val processorOp = ProcessorOp(processor.getName, parallism)
+    dag.addVertex(processorOp)
+    dag.addEdge(root, Shuffle, processorOp)
```

Comment:
Must a process use Shuffle edge? Maybe we can ask Processor to specify whether it preserves partition info or not?

A more general question: what's the criteria for Direct and Shuffle? I assume it is similar to the concepts in Spark. But for Spark, as it using KV, it is simple:
1. Direct means no repartition is needed as the key, if using the same partitioner, will result in the same partition.
2. Otherwise Shuffle.

We need a clear definition of them.